sylvia-derive/src/contract/communication/reply.rs

```diff
@@ -1,60 +1,268 @@
+    /// [`function_name`](crate::types::msg_variant::MsgVariant::function_name) not anotated with
+    /// the `#[sv::msg(..)]` attribute with the `handlers` parameter.
+    fn emit_reply_ids(&'a self) -> impl Iterator<Item = TokenStream> + 'a {
+        self.reply_data.iter().enumerate().map(|(id, data)| {
+            let id = id as u64;
+            let reply_id = &data.reply_id;
+            quote! {
+                pub const #reply_id : u64 = #id ;
+            }
+        })
     }
 }
 
 trait ReplyVariants<'a> {
-    fn as_reply_ids(&'a self) -> impl Iterator<Item = Ident> + 'a;
+    /// Maps to [Vec<ReplyData>].
+    /// Validation for duplicates and overlaps should be done in this method.
+    fn as_reply_data(&self) -> Vec<ReplyData>;
 }
 
 impl<'a> ReplyVariants<'a> for MsgVariants<'a, GenericParam> {
-    fn as_reply_ids(&'a self) -> impl Iterator<Item = Ident> + 'a {
-        self.variants()
-            .flat_map(|variant| {
-                if variant.msg_attr().handlers().is_empty() {
-                    return vec![variant.function_name()];
+    fn as_reply_data(&self) -> Vec<ReplyData> {
+        let mut reply_data: Vec<ReplyData> = vec![];
+
+        self
+            .variants()
+            .flat_map(ReplyVariant::as_reply_data)
+            .for_each(|(reply_id, function_name, reply_on)| {
+                match reply_data.iter_mut().find(|existing_data| existing_data.reply_id == reply_id) {
+                    Some(existing_data)
+                        if existing_data.handlers
+                            .iter()
+                            .any(|(_, prev_reply_on)| {
+                                prev_reply_on.excludes(&reply_on)
+                            }) =>
+                    {
+                        existing_data.handlers.iter().for_each(|(prev_function_name, prev_reply_on)|{
+                            emit_error!(reply_id.span(), "Duplicated reply handler.";
+                                note = prev_function_name.span() => format!("Previous definition of handler={} for reply_on={}", existing_data.reply_id.to_string(), prev_reply_on)
+                            )
+                        })
+                    },
+                    Some(existing_data) => {
+                        existing_data.handlers.push((function_name, reply_on))
+                    },
+                    None => reply_data.push(ReplyData::new (reply_id, function_name, reply_on)),
+               }});
+
+        reply_data
+    }
+}
+
+struct ReplyData<'a> {
+    pub reply_id: Ident,
+    pub handlers: Vec<(&'a Ident, ReplyOn)>,
+}
+
+impl<'a> ReplyData<'a> {
+    pub fn new(reply_id: Ident, function_name: &'a Ident, reply_on: ReplyOn) -> Self {
+        Self {
+            reply_id,
+            handlers: vec![(function_name, reply_on)],
+        }
+    }
+
+    fn emit_match_arms(&self, contract: &Type, generics: &[&GenericParam]) -> TokenStream {
```

Comment:
This method is pretty complex - consider splitting it into few more simpler functions